load-gen/src/test/java/io/smartcat/data/loader/datasource/CSVDataSourceTest.java

```diff
@@ -0,0 +1,232 @@
+        // GIVEN
+        boolean result = false;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(true, result);
+    }
+
+    @Test
+    public void hasNext_should_return_false_at_the_beginning_of_empty_file() throws Exception {
+        // GIVEN
+        boolean result = false;
+        try (CSVParser csvParser = getCsvParser("empty-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(false, result);
+    }
+
+    @Test
+    public void hasNext_should_return_false_at_the_end_of_file() throws Exception {
+        // GIVEN
+        boolean result = true;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+            // file has NUM_OF_RECORDS_IN_SIMPLE_CSV records, run getNext() NUM_OF_RECORDS_IN_SIMPLE_CSV times and then
+            // try hasNext()
+            for (int i = 0; i < NUM_OF_RECORDS_IN_SIMPLE_CSV; i++) {
+                dataSource.getNext(0);
+            }
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(false, result);
+    }
+
+    @Test
+    public void first_getNext_should_return_first_row() throws Exception {
+        // GIVEN
+        User expected = new User("John", "Doe", 31, "555-232-13143", true);
+        User result = null;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.getNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(expected, result);
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void getNext_at_the_end_of_file_should_throw_NoSuchElementException() throws Exception {
+        // GIVEN
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+            // file has NUM_OF_RECORDS_IN_SIMPLE_CSV records, run getNext() NUM_OF_RECORDS_IN_SIMPLE_CSV times and then
+            // try hasNext()
+            for (int i = 0; i < NUM_OF_RECORDS_IN_SIMPLE_CSV; i++) {
+                dataSource.getNext(0);
+            }
+
+            // WHEN
+            dataSource.getNext(0);
+        }
+
+        // THEN
+        // NoSuchElementException is thrown
+    }
+
+    @Test(expected = IllegalStateException.class)
```

Comment:
I always avoid using this except when I create domain exception wrapping some wide-spread exception. This way of doing tests (with expected and class of exception) will hide place of exception. 

Let us say that you have class A using class B and you are testing class A and expecting `IllegalStateException`. You write test like this and class B actually throws `IllegalStateException`. You would think that your test is green but in reality exception is thrown deeper in class tree.

We used on lot of places [AssertJ](http://joel-costigliola.github.io/assertj/) library and if you are using Java8 you can do nice exceptions assert:

```
Throwable thrown = catchThrowable(() -> { throw new Exception("boom!"); });
assertThat(thrown).hasMessageContaining("boom");
```